usr/sbin/p11sak/p11sak.c

```diff
@@ -6254,6 +6607,90 @@ static int p11sak_pem_password_cb(char *buf, int size, int rwflag,
     return len;
 }
 
+/*
+ * Reads an openssl X509 from PEM-type input file. No error msgs here, 
+ * calling routine will issue any errors.
+ */
+static CK_RV p11sak_x509_from_pem_file(X509 **x509)
+{
+    X509 *x = NULL;
+    BIO *bio = NULL;
+    CK_RV rc;
+
+    bio = BIO_new_file(opt_file, "r");
+    if (bio == NULL) {
+        rc = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    x = PEM_read_bio_X509(bio, NULL, p11sak_pem_password_cb, NULL);
+    if (x == NULL) {
+        rc = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    *x509 = x;
+    rc = CKR_OK;
+
+done:
+    if (bio != NULL)
+        BIO_free(bio);
+
+    return rc;
+}
+
+/*
+ * Reads an openssl X509 from binary input file. No error msgs here, 
+ * calling routine will issue any errors.
+ */
+static CK_RV p11sak_x509_from_der_file(X509 **x509)
+{
+    CK_BYTE *value;
+    const CK_BYTE *tmp_value;
+    CK_ULONG value_len;
+    X509 *x;
+    struct stat sb;
+    FILE *fp = NULL;
+    CK_RV rc;
+
+    if (stat(opt_file, &sb) != 0)
+        return CKR_FUNCTION_FAILED;
+
+    value_len = sb.st_size;
+    value = malloc(value_len);
+    if (value == NULL)
+        return CKR_HOST_MEMORY;
+
+    fp = fopen(opt_file, "r");
+    if (fp == NULL) {
+        rc = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    if (fread(value, value_len, 1, fp) != 1) {
+        rc = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    tmp_value = (CK_BYTE *)value;
+    x = d2i_X509(NULL, &tmp_value, value_len);
+    if (x == NULL) {
+        rc = CKR_FUNCTION_FAILED;
```

Comment:
OK, this we want to ignore silently, because we try it as PEM file then.

Reply:
But maybe we want to return a specific return code (e.g. CKR_ARGUMENTS_BAD) here, so that the caller can distinguish between a catastrophic error (where it should stop), or because its not a DER file.